src/tqec/compile/tree/annotators/polygons.py

```diff
@@ -0,0 +1,72 @@
+from typing_extensions import override
+
+from tqec.compile.blocks.layers.atomic.layout import LayoutLayer
+from tqec.compile.tree.annotations import Polygon
+from tqec.compile.tree.node import LayerNode, NodeWalker
+from tqec.utils.exceptions import TQECException
+from tqec.utils.position import Shift2D
+
+
+class AnnotatePolygonOnLayerNode(NodeWalker):
+    def __init__(self, k: int):
+        self._k = k
+
+    @override
+    def visit_node(self, node: LayerNode) -> None:
+        if not node.is_leaf:
+            return
+        assert isinstance(node._layer, LayoutLayer)
+        node.get_annotations(self._k).polygons = generate_polygons_for_layout_layer(
+            node._layer, self._k
+        )
+        node.set_circuit_annotation(self._k, node._layer.to_circuit(self._k))
+
+
+def generate_polygons_for_layout_layer(layer: LayoutLayer, k: int) -> list[Polygon]:
+    template, plaquettes = layer.to_template_and_plaquettes()
+
+    _indices = list(range(1, template.expected_plaquettes_number + 1))
+    template_plaquettes = template.instantiate(k, _indices)
+    increments = template.get_increments()
+
+    polygons: list[Polygon] = []
+
+    for row_index, line in enumerate(template_plaquettes):
+        for column_index, plaquette_index in enumerate(line):
+            if plaquette_index != 0:
+                # Computing the offset that should be applied to each qubits.
+                plaquette = plaquettes[plaquette_index]
+                debug_info = plaquette.debug_information
+                if debug_info is None or debug_info.rpng is None:
+                    raise TQECException(
+                        "Plaquette basis can only be determined for plaquettes with rpng information."
+                    )
```

Comment:
Should we raise here, or maybe just output a polygon of a different colour (e.g., yellow) to allow the user to still have the polygons around? I prefer the second option, else any computation involving stretched stabilizer will raise, reducing the usefulness of annotating polygon.